### DISCUSSION TITLE
32 bit exceptions not working correctly in msys2 bash

### DISCUSSION THREAD

**Discussion by revelator** [msys2/MINGW-packages#15232] [Q&A]
This is a bit of a strange one, static exceptions are broken for 32 bit and it seems to be caused by something in the msys2 environment.
Experiment try building a static gcc then do some exception tests in msys2 bash -> result segmentation fault is thrown before it even begins the test, try the same gcc in old msys.org bash and static exceptions are working again huh ?!?. I cannot for the life of me figure this one out except that something seems to go ape when using the msys2 shell. I tried without mintty as well and same problem so it does not seem to be because of that. this one had me baffled for quite a while trying to make a stable port of TDM gcc for use with msys2 except a ton of stuff simply wont work because of the above.  i wonder if the 32 bit msys2 would yield different results ? but i cannot seem to find it anymore to try it out.

**Comment by mmuetzel** [comment]
IIUC, @jeremyd2019 hosts a fork on GitHub that still builds updated packages for 32-bit:
https://github.com/jeremyd2019/msys2-build32/releases

**Reply by revelator** [reply]
thanks :) ill get on it pronto.

**Reply by jeremyd2019** [reply]
32-bit is still "unsupported".  It might be helpful to you to use a recent installer https://github.com/jeremyd2019/msys2-installer/releases.  I skipped building a 32-bit "release" corresponding to the latest 64-bit one because I had some merging to do in that repo, and because the main update was msys2-runtime (which needs to stay on 3.3 for 32-bit support because cygwin ripped out all 32-bit support in 3.4).  The next release there will use the msys2-runtime-3.3 package instead of the msys2-runtime package because that got updated after msys2-runtime updated to 3.4.

**Reply by revelator** [reply]
thanks that should help :)

**Reply by revelator** [reply]
completed a build using the old msys.org shell, if anyone could try it out id be happy to hear of the results.
the TDM gcc compiler is mostly intended for standalone builds by cutting down on the gcc dll runtimes needed while still maintaining exception support. it does this by using an atomic table to hold the exception data in memory (SEH builds dont need this hack as the exception model is native to windows).  Old msys has no package manager so the compilers are just zipped up with 7z. you can just unzip them to the Msys2 directory but since they are named Tdm32 and Tdm64 you will need to add a few new targets in /etc/msystem and /etc/profile otherwise you could just rename them to mingw32 and mingw64 and you should be good to go (might need to change the compiler prefixes in the pkgconfig files from /tdm* to /mingw* as well).

so far in msys2 i have built all packages except rust because rust depends on libgcc_eh.a and the TDM compilers dont use it (TDM builds as if the compiler itself is a statically built gcc so the unwinders are added to libgcc.a instead). It is however possible to use the dll runtimes if wanted (in case some project croaks on the static runtimes) by adding -shared-libgcc and or -shared-libstdc++ to LDFLAGS.

https://sourceforge.net/projects/cbadvanced/files/TDM%20Testbuild%20old%20Msys/

**Reply by revelator** [reply]
still investigating, had to modify the 32 bit msys2 environment a bit to accept the tdm* prefix so i used some time on that.
then i need to rebuild gcc binutils and then the stuff that is borked on the 64 bit version so it will probably take a while.
tested the existing packages but even though they behave a bit better the exceptions are still broken for some of the tools that have been linked with the static runtimes. i allready did build a few of the packages on old msys 1.0 and they work when compiled with that, but the old msys runtime handles stack memory rather badly so id prefer to use the newer msys2.

**Reply by revelator** [reply]
turns out my old environment had become broken due to the many name changes in pacmans repo (pacman has some issues cleaning up at times i noticed...) the update to the new msys2 64 bit only runtime finally broke the camels back and nothing worked anymore sigh.
had to reinstall msys2 from scratch and go from there, as it turns out after i did that everything just worked again.
I can also report that my TDM compiler fork now works 100% omgawd!!! :D we can close this now.

**Reply by revelator** [reply]
well heres a blast from the past xD i modified a few things and now i have an msys2 shell with full mingw.org compatibility (yeah the really old mingw way before mingw-w64-etc) ;) actually quite a few packages from the repo builds as mingw-pc-i686 packages despite the api being rather out of date. the upside to those who still dabble in win9x is that they can now have access to stuff that was at best a pipe dream before. heh i even got python to build using the old api Oo took a few hacks but it works.
